fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java

```diff
@@ -7257,4 +7259,46 @@ public static boolean isStoredTableNamesLowerCase() {
     public static boolean isTableNamesCaseInsensitive() {
         return GlobalVariable.lowerCaseTableNames == 2;
     }
+
+    public void compactTable(AdminCompactTableStmt stmt) throws DdlException {
+        String dbName = stmt.getDbName();
+        String tableName = stmt.getTblName();
+
+        String type = stmt.getCompactionType();
+        if (type == null || (!type.equals("base") && !type.equals("cumulative"))) {
+            throw new DdlException("compaction type should be [BASE] or [CUMULATIVE]");
+        }
+
+        Database db = this.getDbOrDdlException(dbName);
+        OlapTable olapTable = db.getOlapTableOrDdlException(tableName);
+
+        olapTable.writeLock();
+        try {
+            AgentBatchTask batchTask = new AgentBatchTask();
+            List<String> partitionNames = stmt.getPartitions();
+            LOG.info("Table compaction. database: {}, table: {}, partition: {}, type: {}", dbName, tableName,
+                    Joiner.on(", ").join(partitionNames), type);
+            for (String parName : partitionNames) {
+                Partition partition = olapTable.getPartition(parName);
+                if (partition == null) {
+                    throw new DdlException("partition[" + parName + "] not exist in table[" + tableName + "]");
+                }
+
+                for (MaterializedIndex idx : partition.getMaterializedIndices(IndexExtState.VISIBLE)) {
+                    for (Tablet tablet : idx.getTablets()) {
+                        for (Replica replica : tablet.getReplicas()) {
+                            CompactionTask compactionTask = new CompactionTask(replica.getBackendId(), db.getId(),
+                                    olapTable.getId(), partition.getId(), idx.getId(), tablet.getId(),
+                                    olapTable.getSchemaHashByIndexId(idx.getId()), type, 5000);
+                            batchTask.addTask(compactionTask);
+                        }
+                    }
+                } // indices
+            }
+            // send task immediately
+            AgentTaskExecutor.submit(batchTask);
```

Comment:
submit task outside the lock

Reply:
Timeout for 5000 seconds?
I think the timeout can be calculated by BE.

Reply:
> submit task outside the lock

Done.